pkg/templater/funcmap.go

```diff
@@ -0,0 +1,42 @@
+package templater
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"text/template"
+)
+
+var ErrBadFuncRet = errors.New("wrong number of return values")
+
+// WrapFuncMapWithData wraps a funcMap to a FuncMap that passes data as a
+// first element when calling each element.
+func WrapFuncMapWithData(funcMap template.FuncMap, data interface{}) template.FuncMap {
```

Comment:
Trying to figure out why is it necessary.
The output of this function is an array of functions (FuncMap) that contains functions that hold some data (that would be the controlled data in the implementation), and that appends that "hold-data" to the given argument-data, then calls the underlying function on the total-data.
So a number of questions:
1. Is my understanding correct?
2. Why not initialize the template structure with the original FuncMap, and pass the total-data (controlled + uncontrolled) to Execute()?
3. If it's still necessary to wrap the functions, can we wrap those functions and values in functions instead of reflections to make it more understandable (and probably maintainable)?

By "wrapping with functions" I mean something like:
```go
ret[k] = func(args ...interface{}) (interface{}, error) {
    argVals := make([]interface{}, 0, len(args)+1)
    argVals = append(argVals, data)
    argVals = append(argVals, args...)
    return v(argVals)
}
```
(holding the controlled data and the underlying function)

Reply:
By points...

(1) Yes, this is what this function does.

(2) Execute() is somewhat broken: I could find no mechanism to pass side-data into functions.  The 2nd arg `data` does *not* work for this, because any exported members of it are visible not only to functions but also to the template being expanded.  And I am afraid of that access in templates, I have less control over them and they are a DSL that is difficult to analyze.  In particular note the template function `call`, and consider how much damage a user allowed to create new templates will be able to do with the auth service, if that is supplied on `data`.

(3) I am not sure why wrapping functions manually is less error-prone than performing it automatically.  I prefer to pay the added price of having to understand reflection _now_ but **once**, to having to check each definition of a function wrapper.  Also I don't think that your example will compile without redefining `v` to take all arguments as `interface{}`.  The way to do it would probably be to specify explicitly the types of all arguments to `v`, and just copy the argument list into the call to `v`.